File: CSharp/Concentus/Silk/CorrelateMatrix.cs
/* Copyright (c) 2006-2011 Skype Limited. All Rights Reserved
   Ported to C# by Logan Stromberg

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

   - Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

   - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

   - Neither the name of Internet Society, IETF or IETF Trust, nor the
   names of specific contributors, may be used to endorse or promote
   products derived from this software without specific prior written
   permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

namespace Concentus.Silk
{
    using Concentus.Common;
    using Concentus.Common.CPlusPlus;
    using Concentus.Silk.Enums;
    using Concentus.Silk.Structs;
    using System.Diagnostics;

    /**********************************************************************
 * Correlation Matrix Computations for LS estimate.
 **********************************************************************/
    internal static class CorrelateMatrix
    {
        /* Calculates correlation vector X'*t */
        internal static void silk_corrVector(
            Pointer<short> x,                                     /* I    x vector [L + order - 1] used to form data matrix X                         */
            Pointer<short> t,                                     /* I    Target vector [L]                                                           */
            int L,                                      /* I    Length of vectors                                                           */
            int order,                                  /* I    Max lag for correlation                                                     */
            Pointer<int> Xt,                                    /* O    Pointer to X'*t correlation vector [order]                                  */
            int rshifts                                /* I    Right shifts of correlations                                                */
        )
        {
            int lag, i;
            Pointer<short> ptr1, ptr2;
            int inner_prod;

            ptr1 = x.Point(order - 1); /* Points to first sample of column 0 of X: X[:,0] */
            ptr2 = t;
            /* Calculate X'*t */
            if (rshifts > 0)
            {
                /* Right shifting used */
                for (lag = 0; lag < order; lag++)
                {
                    inner_prod = 0;
                    for (i = 0; i < L; i++)
                    {
                        inner_prod += Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[i], ptr2[i]), rshifts);
                    }
                    Xt[lag] = inner_prod; /* X[:,lag]'*t */
                    ptr1 = ptr1.Point(-1); /* Go to next column of X */
                }
            }
            else {
                Inlines.OpusAssert(rshifts == 0);
                for (lag = 0; lag < order; lag++)
                {
                    Xt[lag] = Inlines.silk_inner_prod_aligned(ptr1, ptr2, L); /* X[:,lag]'*t */
                    ptr1 = ptr1.Point(-1); /* Go to next column of X */
                }
            }
        }

        /* Calculates correlation matrix X'*X */
        internal static void silk_corrMatrix(
            Pointer<short> x,                                     /* I    x vector [L + order - 1] used to form data matrix X                         */
            int L,                                      /* I    Length of vectors                                                           */
            int order,                                  /* I    Max lag for correlation                                                     */
            int head_room,                              /* I    Desired headroom                                                            */
            Pointer<int> XX,                                    /* O    Pointer to X'*X correlation matrix [ order x order ]                        */
            BoxedValue<int> rshifts                               /* I/O  Right shifts of correlations                                                */
        )
        {
            int i, j, lag, head_room_rshifts;
            int energy, rshifts_local;
            Pointer<short> ptr1, ptr2;

            BoxedValue<int> boxed_energy = new BoxedValue<int>();
            BoxedValue<int> boxed_rshift = new BoxedValue<int>();
            /* Calculate energy to find shift used to fit in 32 bits */
            SumSqrShift.silk_sum_sqr_shift(boxed_energy, boxed_rshift, x, L + order - 1);
            energy = boxed_energy.Val;
            rshifts_local = boxed_rshift.Val;
            /* Add shifts to get the desired head room */
            head_room_rshifts = Inlines.silk_max(head_room - Inlines.silk_CLZ32(energy), 0);

            energy = Inlines.silk_RSHIFT32(energy, head_room_rshifts);
            rshifts_local += head_room_rshifts;

            /* Calculate energy of first column (0) of X: X[:,0]'*X[:,0] */
            /* Remove contribution of first order - 1 samples */
            for (i = 0; i < order - 1; i++)
            {
                energy -= Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(x[i], x[i]), rshifts_local);
            }
            if (rshifts_local < rshifts.Val)
            {
                /* Adjust energy */
                energy = Inlines.silk_RSHIFT32(energy, rshifts.Val - rshifts_local);
                rshifts_local = rshifts.Val;
            }

            /* Calculate energy of remaining columns of X: X[:,j]'*X[:,j] */
            /* Fill out the diagonal of the correlation matrix */
            Inlines.matrix_adr(XX, 0, 0, order)[0] = energy;
            ptr1 = x.Point(order - 1); /* First sample of column 0 of X */
            for (j = 1; j < order; j++)
            {
                energy = Inlines.silk_SUB32(energy, Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[L - j], ptr1[L - j]), rshifts_local));
                energy = Inlines.silk_ADD32(energy, Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[-j], ptr1[-j]), rshifts_local));
                Inlines.matrix_adr(XX, j, j, order)[0] = energy;
            }

            ptr2 = x.Point(order - 2); /* First sample of column 1 of X */
                                  /* Calculate the remaining elements of the correlation matrix */
            if (rshifts_local > 0)
            {
                /* Right shifting used */
                for (lag = 1; lag < order; lag++)
                {
                    /* Inner product of column 0 and column lag: X[:,0]'*X[:,lag] */
                    energy = 0;
                    for (i = 0; i < L; i++)
                    {
                        energy += Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[i], ptr2[i]), rshifts_local);
                    }
                    /* Calculate remaining off diagonal: X[:,j]'*X[:,j + lag] */
                    Inlines.matrix_adr(XX, lag, 0, order)[0] = energy;
                    Inlines.matrix_adr(XX, 0, lag, order)[0] = energy;
                    for (j = 1; j < (order - lag); j++)
                    {
                        energy = Inlines.silk_SUB32(energy, Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[L - j], ptr2[L - j]), rshifts_local));
                        energy = Inlines.silk_ADD32(energy, Inlines.silk_RSHIFT32(Inlines.silk_SMULBB(ptr1[-j], ptr2[-j]), rshifts_local));
                        Inlines.matrix_adr(XX, lag + j, j, order)[0] = energy;
                        Inlines.matrix_adr(XX, j, lag + j, order)[0] = energy;
                    }
                    ptr2 = ptr2.Point(-1); /* Update pointer to first sample of next column (lag) in X */
                }
            }
            else {
                for (lag = 1; lag < order; lag++)
                {
                    /* Inner product of column 0 and column lag: X[:,0]'*X[:,lag] */
                    energy = Inlines.silk_inner_prod_aligned(ptr1, ptr2, L);
                    Inlines.matrix_adr(XX, lag, 0, order)[0] = energy;
                    Inlines.matrix_adr(XX, 0, lag, order)[0] = energy;
                    /* Calculate remaining off diagonal: X[:,j]'*X[:,j + lag] */
                    for (j = 1; j < (order - lag); j++)
                    {
                        energy = Inlines.silk_SUB32(energy, Inlines.silk_SMULBB(ptr1[L - j], ptr2[L - j]));
                        energy = Inlines.silk_SMLABB(energy, ptr1[-j], ptr2[-j]);
                        Inlines.matrix_adr(XX, lag + j, j, order)[0] = energy;
                        Inlines.matrix_adr(XX, j, lag + j, order)[0] = energy;
                    }
                    ptr2 = ptr2.Point(-1);/* Update pointer to first sample of next column (lag) in X */
                }
            }
            rshifts.Val = rshifts_local;
        }

    }
}

File: Go/opus/CorrelateMatrix.go
package opus

type correlateMatrix struct{}

var CorrelateMatrix correlateMatrix

func (c *correlateMatrix) silk_corrVector(
	x []int16,
	x_ptr int,
	t []int16,
	t_ptr int,
	L int,
	order int,
	Xt []int,
	rshifts int) {

	var lag, i int
	var ptr1, ptr2 int
	var inner_prod int

	ptr1 = x_ptr + order - 1
	ptr2 = t_ptr

	if rshifts > 0 {
		for lag = 0; lag < order; lag++ {
			inner_prod = 0
			for i = 0; i < L; i++ {
				inner_prod += silk_RSHIFT32(silk_SMULBB(int(x[ptr1+i]), int(t[ptr2+i])), rshifts)
			}
			Xt[lag] = inner_prod
			ptr1--
		}
	} else {
		OpusAssert(rshifts == 0)
		for lag = 0; lag < order; lag++ {
			Xt[lag] = silk_inner_prod(x, ptr1, t, ptr2, L)
			ptr1--
		}
	}
}

func (c *correlateMatrix) silk_corrMatrix(
	x []int16,
	x_ptr int,
	L int,
	order int,
	head_room int,
	XX []int,
	XX_ptr int,
	rshifts *BoxedValueInt) {

	var i, j, lag, head_room_rshifts int
	var energy, rshifts_local int

	boxed_energy := BoxedValueInt{Val: 0}
	boxed_rshifts_local := BoxedValueInt{Val: 0}
	silk_sum_sqr_shift5(&boxed_energy, &boxed_rshifts_local, x, x_ptr, L+order-1)
	energy = boxed_energy.Val
	rshifts_local = boxed_rshifts_local.Val

	head_room_rshifts = silk_max(head_room-silk_CLZ32(energy), 0)

	energy = silk_RSHIFT32(energy, head_room_rshifts)
	rshifts_local += head_room_rshifts

	for i = x_ptr; i < x_ptr+order-1; i++ {
		energy -= silk_RSHIFT32(silk_SMULBB(int(x[i]), int(x[i])), rshifts_local)
	}
	if rshifts_local < rshifts.Val {
		energy = silk_RSHIFT32(energy, rshifts.Val-rshifts_local)
		rshifts_local = rshifts.Val
	}

	MatrixSet(XX, XX_ptr, 0, 0, order, energy)
	ptr1 := x_ptr + order - 1
	for j = 1; j < order; j++ {
		energy = silk_SUB32(energy, silk_RSHIFT32(silk_SMULBB(int(x[ptr1+L-j]), int(x[ptr1+L-j])), rshifts_local))
		energy = silk_ADD32(energy, silk_RSHIFT32(silk_SMULBB(int(x[ptr1-j]), int(x[ptr1-j])), rshifts_local))
		MatrixSet(XX, XX_ptr, j, j, order, energy)
	}

	ptr2 := x_ptr + order - 2
	if rshifts_local > 0 {
		for lag = 1; lag < order; lag++ {
			energy = 0
			for i = 0; i < L; i++ {
				energy += silk_RSHIFT32(silk_SMULBB(int(x[ptr1+i]), int(x[ptr2+i])), rshifts_local)
			}
			MatrixSet(XX, XX_ptr, lag, 0, order, energy)
			MatrixSet(XX, XX_ptr, 0, lag, order, energy)
			for j = 1; j < (order - lag); j++ {
				energy = silk_SUB32(energy, silk_RSHIFT32(silk_SMULBB(int(x[ptr1+L-j]), int(x[ptr2+L-j])), rshifts_local))
				energy = silk_ADD32(energy, silk_RSHIFT32(silk_SMULBB(int(x[ptr1-j]), int(x[ptr2-j])), rshifts_local))
				MatrixSet(XX, XX_ptr, lag+j, j, order, energy)
				MatrixSet(XX, XX_ptr, j, lag+j, order, energy)
			}
			ptr2--
		}
	} else {
		for lag = 1; lag < order; lag++ {
			energy = silk_inner_prod(x, ptr1, x, ptr2, L)
			MatrixSet(XX, XX_ptr, lag, 0, order, energy)
			MatrixSet(XX, XX_ptr, 0, lag, order, energy)
			for j = 1; j < (order - lag); j++ {
				energy = silk_SUB32(energy, silk_SMULBB(int(x[ptr1+L-j]), int(x[ptr2+L-j])))
				energy = silk_SMLABB(energy, int(x[ptr1-j]), int(x[ptr2-j]))
				MatrixSet(XX, XX_ptr, lag+j, j, order, energy)
				MatrixSet(XX, XX_ptr, j, lag+j, order, energy)
			}
			ptr2--
		}
	}
	rshifts.Val = rshifts_local
}
